amgcl/backend/builtin.hpp

```diff
@@ -605,6 +606,44 @@ struct inner_product_impl<
 };
 
 template < class Vec1, class Vec2 >
+struct inner_product_impl<
+    Vec1, Vec2,
+    typename boost::enable_if<
+            typename boost::mpl::and_<
+                typename is_builtin_vector<Vec1>::type,
+                typename is_builtin_vector<Vec2>::type,
+                typename boost::is_complex<typename value_type<Vec1>::type>::type
+                >::type
+        >::type
+    >
+{
+    typedef typename value_type<Vec1>::type V;
+
+    static V get(const Vec1 &x, const Vec2 &y)
+    {
+        const size_t n = x.size();
+        V sum = 0;
+
+        {
+            size_t chunk_start = 0;
+            size_t chunk_end   = n;
+
+            V s = 0;
+            V c = 0;
+            for(size_t i = chunk_start; i < chunk_end; ++i) {
+                V d = x[i] * std::conj(y[i]) - c;
```

Comment:
We could also introduce `conj_transp` operation that would be a noop by default, `std::conj` for complex numbers, and actual conjugate transpose for generic nonscalar values in the future. This way we could keep single implementation of inner product.